src/main/java/com/wavefront/opentracing/reporting/WavefrontSpanReporter.java

```diff
@@ -54,24 +60,62 @@ public Builder withSource(String source) {
       return this;
     }
 
+    /**
+     * Set max queue size of in-memory buffer. Incoming spans are dropped if buffer is full.
+     *
+     * @param maxQueueSize Max queue size of in-memory buffer
+     * @return {@code this}
+     */
+    public Builder withMaxQueueSize(int maxQueueSize) {
+      this.maxQueueSize = maxQueueSize;
+      return this;
+    }
+
     /**
      * Builds a {@link WavefrontSpanReporter} for sending opentracing spans to a
      * WavefrontSender that can send those spans either be a via proxy or direct ingestion.
      *
      * @return {@link WavefrontSpanReporter}
      */
     public WavefrontSpanReporter build(WavefrontSender wavefrontSender) {
-      return new WavefrontSpanReporter(wavefrontSender, this.source);
+      return new WavefrontSpanReporter(wavefrontSender, this.source, this.maxQueueSize);
     }
   }
 
-  private WavefrontSpanReporter(WavefrontSender wavefrontSender, String source) {
+  private WavefrontSpanReporter(WavefrontSender wavefrontSender, String source, int maxQueueSize) {
     this.wavefrontSender = wavefrontSender;
     this.source = source;
+    this.spanBuffer = new LinkedBlockingQueue<>(maxQueueSize);
+
+    sendingThread = new Thread(this, "wavefrontSpanReporter");
+    sendingThread.setDaemon(true);
+    sendingThread.start();
+  }
+
+  @Override
+  public void run() {
+    while (!stop) {
+      try {
+        WavefrontSpan span = spanBuffer.take();
+        send(span);
+      } catch (InterruptedException ex) {
+        if (logger.isLoggable(Level.INFO)) {
+          logger.info("reporting thread interrupted");
+        }
+      } catch (Exception ex) {
+        logger.log(Level.WARNING, "Error processing buffer", ex);
+      }
+    }
   }
 
   @Override
   public void report(WavefrontSpan span) {
+    if (!spanBuffer.offer(span)) {
+      logger.warning("Dropping span: " + span);
```

Comment:
Not a blocker but in the future we should consider discarding older spans to favor more recent ones.